span_compressed.go

```diff
@@ -0,0 +1,321 @@
+			count:               1,
+			sum:                 int64(s.Duration),
+			compressionStrategy: strategy,
+		}
+	}
+
+	s.composite.count++
+	s.composite.sum += int64(sibling.Duration)
+	siblingTimestamp := sibling.timestamp.Add(sibling.Duration)
+	if siblingTimestamp.After(s.composite.lastSiblingEndTime) {
+		s.composite.lastSiblingEndTime = siblingTimestamp
+	}
+	return true
+}
+
+func (s *Span) canCompressStandard(sibling *Span) int {
+	if !s.isSameKind(sibling) {
+		return 0
+	}
+
+	if s.isExactMatch(sibling) {
+		if s.durationLowerOrEq(sibling,
+			s.tx.compressedSpans.exactMatchMaxDuration,
+		) {
+			return compressedStrategyExactMatch
+		}
+		return 0
+	}
+	if s.isSameKind(sibling) {
+		if s.durationLowerOrEq(sibling,
+			s.tx.compressedSpans.sameKindMaxDuration,
+		) {
+			s.Name = "Calls to " + s.Context.destinationService.Resource
+			return compressedStrategyExactMatch
+		}
+	}
+	return 0
+}
+
+func (s *Span) canCompressComposite(sibling *Span) int {
+	if s.composite.empty() {
+		return 0
+	}
+	switch s.composite.compressionStrategy {
+	case compressedStrategyExactMatch:
+		if s.isExactMatch(sibling) && s.durationLowerOrEq(sibling,
+			s.tx.compressedSpans.exactMatchMaxDuration,
+		) {
+			return compressedStrategyExactMatch
+		}
+	case compressedStrategySameKind:
+		if s.isSameKind(sibling) && s.durationLowerOrEq(sibling,
+			s.tx.compressedSpans.sameKindMaxDuration,
+		) {
+			return compressedStrategySameKind
+		}
+	}
+	return 0
+}
+
+func (s *Span) durationLowerOrEq(sibling *Span, max time.Duration) bool {
+	return s.Duration <= max && sibling.Duration <= max
+}
+
+// spanBuffer acts as an intermediary buffer that stores compressable spans on
+// their parents to compress with other compression eligible siblings.
+//
+// Not concurrently safe.
+type spanBuffer struct {
+	span *Span
+}
+
+func (b *spanBuffer) empty() bool { return b.span == nil }
+
+func (b *spanBuffer) store(s *Span) { preserveSpanData(s, b) }
+
+// flush enqueues a span to the tracer event queue, but first, it reports the
+// parent and tx timers when the breakdown metrics are enabled and when it's
+// a compressed span, the duration is adjusted to end when the last sibling
+// ended - first span event timestamp.
+func (b *spanBuffer) flush(s *Span) {
+	// When the span is a composite span, we need to adjust the duration
+	// just before it is reported and no more spans will be compressed into
+	// the composite. If we did this any time before, the duration of the span
+	// would potentially grow over the compressable threshold and result in
+	// compressable span not being compressed and reported separately.
+	if !b.span.composite.empty() {
+		b.span.Duration = b.span.composite.lastSiblingEndTime.Sub(b.span.timestamp)
+	}
+	if !b.span.tx.ended() && b.span.tx.breakdownMetricsEnabled {
+		b.span.reportSelfTimeLockless(b.span.timestamp.Add(b.span.Duration))
+	}
+
+	b.span.enqueue()
+	b.span = s
+	if s != nil {
+		preserveSpanData(s, b)
+	}
```

Comment:
why do we want to store s in b here? after flushing, shouldn't the buffer be emptied?